source/material_model/viscoelastic_plastic.cc

```diff
@@ -77,62 +79,55 @@ namespace aspect
-              std::vector<double> viscosities_viscoplastic(volume_fractions.size());
-              std::vector<double> viscosities_viscoelastic(volume_fractions.size());
+              std::vector<double> viscosities_ve(volume_fractions.size());
+              std::vector<double> viscosities_vep(volume_fractions.size());
+
+              SymmetricTensor<2,dim> stress_old;
+              for (unsigned int j=0; j < SymmetricTensor<2,dim>::n_independent_components; ++j)
+                stress_old[SymmetricTensor<2,dim>::unrolled_to_component_indices(j)] = composition[j];
 
               // Loop through all compositions
               for (unsigned int j=0; j < volume_fractions.size(); ++j)
                 {
 
-                  // Note that edot_ii is the full computed strain rate, which includes elastic stresses
-                  stresses_viscous[j] = 2. * viscosities_pre_yield[j] * edot_ii;
+                  // Calculate viscoelastic viscosity (equation 28 from Moresi et al., 2003)
+                  viscosities_ve[j] = viscosities_pre_yield[j] * dte / (dte + (viscosities_pre_yield[j]/elastic_shear_moduli[j]));
 
+                  // Calculate updated viscoselastic stress magnitude, which will be compared with the plastic yield stress
+                  stresses_ve[j] = viscosities_ve[j] * std::sqrt(std::fabs(second_invariant((2. * (deviator(strain_rate)) + stress_old / (elastic_shear_moduli[j] * dte) ) ) ) );
+
+                  // Calculate plastic yield stress
                   stresses_yield[j] = ( (dim==3)
                                         ?
                                         ( 6.0 * coh[j] * std::cos(phi[j]) + 6.0 * std::max(pressure,0.0) * std::sin(phi[j]) )
                                         / ( std::sqrt(3.0) * (3.0 + std::sin(phi[j]) ) )
                                         :
                                         coh[j] * std::cos(phi[j]) + std::max(pressure,0.0) * std::sin(phi[j]) );
 
-                  // If the viscous stress is greater than the yield strength, rescale the viscosity back to yield surface.
-                  // If the viscous stress is less than the yield stress, the yield viscosity is equal to the pre-yield value.
-                  if ( stresses_viscous[j] >= stresses_yield[j]  )
+                  // If the viscoelastic stress is greater than the yield strength, rescale the viscosity back to yield surface.
+                  // If the viscoelastic stress is less than the yield stress, the yield viscosity is equal to the pre-yield (viscoelastic) value.
+                  if ( stresses_ve[j] >= stresses_yield[j]  )
                     {
-                      viscosities_viscoplastic[j] = stresses_yield[j] / (2.0 * edot_ii);
+                      viscosities_vep[j] = stresses_yield[j] / std::sqrt(std::fabs(second_invariant((2. * (deviator(strain_rate)) + stress_old / (elastic_shear_moduli[j] * dte) ) ) ) );
```

Comment:
You divide by the strain rate invariant here, but you removed the security measure above that replaces the strain rate with a reference strain rate if it is 0. Is this no longer necessary, because if the strain rate is 0 the stresses_ve can never be larger than the stresses_yield? Otherwise we might run into an divide by zero here.

Reply:
Yes, this security measure absolutely should still be there. I'm surprised we did not get a crash on the first nonlinear iteration when the elastic stresses are 0  (time step 0). The code has been adjusted accordingly.